// themr/lib/bindings/react/themrProvider.js
/**
 * @license
 * Copyright (C) 2016 Chi Vinh Le and contributors.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
import * as React from "react";
var ThemrProvider = (function (_super) {
    __extends(ThemrProvider, _super);
    function ThemrProvider(props) {
        var _this = _super.call(this, props) || this;
        _this.vars = props.vars || {};
        _this.transformer = props.transformer;
        return _this;
    }
    ThemrProvider.prototype.componentWillReceiveProps = function (nextProps) {
        this.vars = nextProps.vars || {};
        this.transformer = nextProps.transformer;
    };
    ThemrProvider.prototype.getChildContext = function () {
        return {
            themer: {
                vars: this.vars,
                transformer: this.transformer,
            },
        };
    };
    ThemrProvider.prototype.render = function () {
        return React.Children.only(this.props.children);
    };
    return ThemrProvider;
}(React.Component));
export { ThemrProvider };
ThemrProvider.childContextTypes = {
    themer: React.PropTypes.object,
};
